File: LeetCode/2751-Robot-Collisions.cpp
// 2751. Robot Collisions

class Solution {
public:
    vector<int> survivedRobotsHealths(vector<int>& positions, vector<int>& healths, string directions) {
        int n = positions.size();
        vector<vector<int>> v(n, vector<int>(4));
        for(int i = 0; i < n; i++) {
            v[i][0] = positions[i];
            v[i][1] = healths[i];
            v[i][2] = (directions[i] == 'L') ? 0 : 1;
            v[i][3] = i;
        }
        sort(v.begin(),v.end());
        
        stack<vector<int>> st;
        for(int i = 0; i < n; i++) {
            st.push(v[i]);

            while(st.size() > 1) {
                vector<int> v2 = st.top();
                st.pop();
                vector<int> v1 = st.top();
                st.pop();

                if(!(v1[2] == 1 && v2[2] == 0)) {
                    st.push(v1);
                    st.push(v2);
                    break;
                }
                
                if(v1[1] > v2[1]) {
                    v1[1]--;
                    st.push(v1);
                }
                else if(v1[1] < v2[1]) {
                    v2[1]--;
                    st.push(v2);
                }
            }
        }

        vector<pair<int,int>> v2;
        while(!st.empty()) {
            v2.push_back({ st.top()[3], st.top()[1] });
            st.pop();
        }
        sort(v2.begin(), v2.end());

        vector<int> ans;
        for(auto i : v2) {
            ans.push_back(i.second);
        }
        return ans;
    }
};

File: LeetCode/2751-Robot-Collisions.java
// 2751. Robot Collisions

class Solution {
    public List<Integer> survivedRobotsHealths(int[] positions, int[] healths, String directions) {
        int n = positions.length;

        List<int[]> lst = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            lst.add(new int[]{positions[i], healths[i], directions.charAt(i) == 'L' ? 0 : 1, i});
        }
        Collections.sort(lst, (a, b) -> Integer.compare(a[0], b[0]));

        Stack<int[]> st = new Stack<>();
        for (int[] a : lst) {
            int h = a[1];
            int dir = a[2];
            int robot = a[3];
            if (dir == 1 || st.isEmpty()) {
                st.add(new int[]{h, dir, robot});
            }
            else {
                while (!st.isEmpty() && st.peek()[1] == 1 && st.peek()[0] < h) {
                    st.pop();
                    h--;
                }
                if (!st.isEmpty() && st.peek()[1] == 1) {
                    if (st.peek()[0] == h) {
                        st.pop();
                    }
                    else {
                        st.peek()[0]--;
                    }
                    h = 0;
                }
                if (h > 0) {
                    st.add(new int[]{h, dir, robot});
                }
            }
        }

        List<int[]> tmp = new ArrayList<>(st);
        Collections.sort(tmp, (a, b) -> Integer.compare(a[2], b[2]));

        List<Integer> ans = new ArrayList<>();
        for (int[] a : tmp) {
            ans.add(a[0]);
        }

        return ans;
    }
}
